File: app/schemas/payment.py
from datetime import date
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field


# ─── Customer Payment Schemas ───


class CustomerPaymentBase(BaseModel):

    customer_id: int = Field(
        ...,
        gt=0
    )

    amount: Decimal = Field(
        ...,
        gt=0
    )

    payment_mode: str = Field(
        pattern="^(CASH|UPI|CARD|CHEQUE|BANK_TRANSFER)$"
    )

    payment_type: str = Field(
        pattern="^(ADVANCE|BILL_PAYMENT)$"
    )

    reference_number: str | None = Field(
        default=None,
        max_length=50
    )

    bill_id: int | None = Field(
        default=None,
        gt=0
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class CustomerPaymentCreate(CustomerPaymentBase):
    pass


class CustomerPaymentUpdate(BaseModel):

    amount: Decimal | None = Field(
        default=None,
        gt=0
    )

    payment_mode: str | None = Field(
        default=None,
        pattern="^(CASH|UPI|CARD|CHEQUE|BANK_TRANSFER)$"
    )

    reference_number: str | None = Field(
        default=None,
        max_length=50
    )

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class CustomerPaymentResponse(CustomerPaymentBase):

    id: int

    payment_date: datetime

    collected_by: int | None

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class CustomerPaymentListResponse(BaseModel):

    id: int

    customer_id: int

    customer_code: str

    customer_name: str

    payment_date: datetime

    amount: Decimal

    payment_mode: str

    payment_type: str

    reference_number: str | None

    bill_id: int | None

    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


# ─── Customer Bill Schemas ───


class CustomerBillItemResponse(BaseModel):

    id: int

    milk_type_id: int

    milk_name: str

    quantity: int

    unit_price: Decimal

    amount: Decimal

    model_config = ConfigDict(
        from_attributes=True
    )


class CustomerBillBase(BaseModel):

    customer_id: int = Field(
        ...,
        gt=0
    )

    bill_period_start: date

    bill_period_end: date

    due_date: date | None = None

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class BillGenerateRequest(BaseModel):

    customer_id: int = Field(
        ...,
        gt=0
    )

    bill_period_start: date

    bill_period_end: date

    due_date: date | None = None

    remarks: str | None = Field(
        default=None,
        max_length=255
    )


class CustomerBillResponse(BaseModel):

    id: int

    customer_id: int

    bill_date: date

    bill_period_start: date

    bill_period_end: date

    total_amount: Decimal

    paid_amount: Decimal

    balance_amount: Decimal

    status: str

    due_date: date | None

    remarks: str | None

    is_active: bool

    created_at: datetime

    updated_at: datetime

    items: list[CustomerBillItemResponse]

    model_config = ConfigDict(
        from_attributes=True
    )


class CustomerBillListResponse(BaseModel):

    id: int

    customer_id: int

    customer_code: str

    customer_name: str

    bill_date: date

    bill_period_start: date

    bill_period_end: date

    total_amount: Decimal

    paid_amount: Decimal

    balance_amount: Decimal

    status: str

    due_date: date | None

    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


class BillStatusUpdate(BaseModel):

    status: str = Field(
        pattern="^(PENDING|PARTIAL|PAID|OVERDUE|CANCELLED)$"
    )


# ─── Outstanding Balance Schema ───


class OutstandingBalanceResponse(BaseModel):

    customer_id: int

    customer_code: str

    customer_name: str

    total_billed: Decimal

    total_paid: Decimal

    balance: Decimal

    last_bill_date: date | None

    last_payment_date: datetime | None

File: frontend/src/types/payment.ts
export type PaymentMode = "CASH" | "UPI" | "CARD" | "CHEQUE" | "BANK_TRANSFER";

export type PaymentType = "ADVANCE" | "BILL_PAYMENT";

export type BillStatus = "PENDING" | "PARTIAL" | "PAID" | "OVERDUE" | "CANCELLED";

export interface CustomerPaymentCreate {
  customer_id: number;
  amount: number;
  payment_mode: PaymentMode;
  payment_type: PaymentType;
  reference_number?: string | null;
  bill_id?: number | null;
  remarks?: string | null;
}

export interface CustomerPaymentResponse {
  id: number;
  customer_id: number;
  amount: number;
  payment_mode: string;
  payment_type: string;
  reference_number: string | null;
  bill_id: number | null;
  payment_date: string;
  collected_by: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CustomerPaymentListResponse {
  id: number;
  customer_id: number;
  customer_code: string;
  customer_name: string;
  payment_date: string;
  amount: number;
  payment_mode: string;
  payment_type: string;
  reference_number: string | null;
  bill_id: number | null;
  is_active: boolean;
}

export interface BillGenerateRequest {
  customer_id: number;
  bill_period_start: string;
  bill_period_end: string;
  due_date?: string | null;
  remarks?: string | null;
}

export interface CustomerBillItemResponse {
  id: number;
  milk_type_id: number;
  milk_name: string;
  quantity: number;
  unit_price: number;
  amount: number;
}

export interface CustomerBillResponse {
  id: number;
  customer_id: number;
  bill_date: string;
  bill_period_start: string;
  bill_period_end: string;
  total_amount: number;
  paid_amount: number;
  balance_amount: number;
  status: string;
  due_date: string | null;
  remarks: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  items: CustomerBillItemResponse[];
}

export interface CustomerBillListResponse {
  id: number;
  customer_id: number;
  customer_code: string;
  customer_name: string;
  bill_date: string;
  bill_period_start: string;
  bill_period_end: string;
  total_amount: number;
  paid_amount: number;
  balance_amount: number;
  status: string;
  due_date: string | null;
  is_active: boolean;
}

export interface OutstandingBalanceResponse {
  customer_id: number;
  customer_code: string;
  customer_name: string;
  total_billed: number;
  total_paid: number;
  balance: number;
  last_bill_date: string | null;
  last_payment_date: string | null;
}

export interface PaymentListParams {
  customer_id?: number;
  payment_mode?: string;
  payment_type?: string;
  from_date?: string;
  to_date?: string;
}

export interface BillListParams {
  customer_id?: number;
  status?: string;
  from_date?: string;
  to_date?: string;
}
